syncapi/routing/messages.go

```diff
@@ -380,17 +386,20 @@ func (r *messagesReq) backfill(fromEventIDs []string, limit int) ([]gomatrixserv
 	}
 
 	for _, p := range txn.PDUs {
-		pdus = append(pdus, p.Headered(gomatrixserverlib.RoomVersionV1))
+		event, e := gomatrixserverlib.NewEventFromUntrustedJSON(p, verRes.RoomVersion)
+		if e != nil {
+			continue
+		}
+		pdus = append(pdus, event.Headered(verRes.RoomVersion))
 	}
 	util.GetLogger(r.ctx).WithField("server", srvToBackfillFrom).WithField("new_events", len(pdus)).Info("Storing new events from backfill")
 
 	// Store the events in the database, while marking them as unfit to show
 	// up in responses to sync requests.
-	for _, pdu := range pdus {
-		headered := pdu.Headered(gomatrixserverlib.RoomVersionV1)
+	for i := range pdus {
 		if _, err = r.db.WriteEvent(
 			r.ctx,
-			&headered,
+			&pdus[i],
```

Comment:
I dislike this because it's not clear what form 'PDU' is:
 - An `Event`
 - A `json.RawMessage`
 - A `HeaderedEvent`

Please rename to be clearer.

Reply:
Named it back to `headered`.